collector/report.go

```diff
@@ -0,0 +1,331 @@
+func (a *ActiveReport) Validate() error {
+	if testNameRegexp.MatchString(a.TestName) != true {
+		return errors.New("invalid \"test_name\" field")
+	}
+	if probeASNRegexp.MatchString(a.ProbeASN) != true {
+		return errors.New("invalid \"probe_asn\" field")
+	}
+	if probeCCRegexp.MatchString(a.ProbeCC) != true {
+		return errors.New("invalid \"probe_cc\" field")
+	}
+	if stringInSlice(a.Format, supportedFormats) != true {
+		return fmt.Errorf("invalid \"format\" field, %s", a.Format)
+	}
+	return nil
+}
+
+// IncomingFilename determines the filename of an incoming report
+func (a *ActiveReport) IncomingFilename() string {
+	if stringInSlice(a.Format, supportedFormats) != true {
+		// Defensive coding
+		panic("an unexpected value for format was found. Bailing...")
+	}
+	return fmt.Sprintf("%s.%s",
+		a.ReportID,
+		a.Format)
+}
+
+// SyncFilename creates a filename in the sync directory of the active report
+func (a *ActiveReport) SyncFilename() (string, error) {
+	// Defensive coding.
+	if err := a.Validate(); err != nil {
+		return "", err
+	}
+
+	return fmt.Sprintf("%s-%s-%s-%s-%s-probe-0.2.0.%s",
+		a.CreationTime.Format(TimestampFormat),
+		a.TestName,
+		a.ReportID,
+		a.ProbeASN,
+		a.ProbeCC,
+		a.Format), nil
+}
+
+// BackendExtra is serverside extra metadata
+type BackendExtra struct {
+	SubmissionTime time.Time `json:"submission_time" bson:"submission_time"`
+	MeasurementID  string    `json:"measurement_id" bson:"measurement_id"`
+	ReportID       string    `json:"report_id" bson:"report_id"`
+}
+
+// MeasurementEntry is the structure of measurements submitted by an OONI Probe client
+type MeasurementEntry struct {
+	// These values are added by the pipeline
+	// BucketDate     string `json:"bucket_date"`
+	// ReportFilename string `json:"report_filename"`
+
+	ID                   string       `json:"id"`
+	ReportID             string       `json:"report_id"`
+	TestName             string       `json:"test_name"`
+	TestVersion          string       `json:"test_version"`
+	MeasurementStartTime string       `json:"measurement_start_time,omitempty"`
+	TestStartTime        string       `json:"test_start_time"` // XXX these should actually be time
+	Annotations          interface{}  `json:"annotations"`
+	BackendExtra         BackendExtra `json:"backend_extra"`
+	BackendVersion       string       `json:"backend_version"`
+	DataFormatVersion    string       `json:"data_format_version"`
+	Input                string       `json:"input"`
+	InputHashes          []string     `json:"input_hashes"`
+	Options              []string     `json:"options"`
+	ProbeASN             string       `json:"probe_asn"`
+	ProbeCC              string       `json:"probe_cc"`
+	ProbeCity            string       `json:"probe_city"`
+	ProbeIP              string       `json:"probe_ip"`
+	SoftwareName         string       `json:"software_name"`
+	SoftwareVersion      string       `json:"software_version"`
+	TestHelpers          interface{}  `json:"test_helpers"`
```

Comment:
This may probably be `map[string]interface{}` right?

Reply:
That could be, though I would rather we keep these more complex fields as opaque `interface{}` to avoid triggering possible parsing errors.

Maybe we should add an extra check in the code (or maybe do some work on the pipeline to understand how many measurement if any at all don't follow this definition).

Reply:
I would like us to start enforcing more data correctness at the collector level. I think a sound strategy here is to have data points coming from OONI versions that we care about the most and then make sure we can parse them.